java-checks/src/main/java/org/sonar/java/checks/security/TriggeringSecuringHelper.java

```diff
@@ -0,0 +1,110 @@
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import static org.sonar.java.model.ExpressionUtils.extractIdentifierSymbol;
+import static org.sonar.java.model.ExpressionUtils.getAssignedSymbol;
+
+/**
+ * Helper class implementing the following behavior:
+ *
+ * 1. Test if a {@link MethodInvocationTree} matches the triggering method matcher given to constructor.
+ * 2. If found, get enclosing method (if any).
+ * 3. Visit the body of the enclosing method, calling {@link #processSecuringMethodInvocation(MethodInvocationTree)} for each method invocation.
+ * 4. At the end of the visit, return true if the variable was secured inside the body of the enclosing method.
+ *
+ * It also makes sure that the triggering and securing method call is made on the same symbol.
+ *
+ */
+public abstract class TriggeringSecuringHelper implements Predicate<MethodInvocationTree> {
+
+  private final MethodMatcher triggeringInvocationMatcher;
+
+  public abstract void resetState();
+
+  /**
+   * The method visitor call this method for all method invocation in the method containing the triggering call.
+   * It typically contributes to define if the triggering call is secured or not.
+   * (see {@link #isSecured()})
+   */
+  public abstract void processSecuringMethodInvocation(MethodInvocationTree mit);
+
+  /**
+   * At the end of the visit of the enclosing method body, return true if all expected securing method call were found.
+   */
+  public abstract boolean isSecured();
+
+  TriggeringSecuringHelper(MethodMatcher triggeringInvocationMatcher) {
+    this.triggeringInvocationMatcher = triggeringInvocationMatcher;
+  }
+
+  @Override
+  public boolean test(MethodInvocationTree mit) {
+    if (triggeringInvocationMatcher.matches(mit)) {
+      MethodTree enclosingMethod = ExpressionUtils.getEnclosingMethod(mit);
+      if (enclosingMethod != null) {
+        Optional<Symbol> assignedSymbol = getAssignedSymbol(mit);
+        if (assignedSymbol.isPresent()) {
+          resetState();
+          enclosingMethod.accept(new MethodVisitor(assignedSymbol.get()));
+          return !isSecured();
+        }
+      }
+    }
+    return false;
+  }
+
+  private class MethodVisitor extends BaseTreeVisitor {
+
+    private Symbol variable;
+
+    private MethodVisitor(Symbol variable) {
+      this.variable = variable;
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree methodInvocation) {
+      if (isSameVariableSymbol(methodInvocation)) {
+        processSecuringMethodInvocation(methodInvocation);
+      }
+      super.visitMethodInvocation(methodInvocation);
+    }
+
+    private boolean isSameVariableSymbol(MethodInvocationTree mit) {
```

Comment:
Isn't this method the same as in: 
https://github.com/SonarSource/sonar-java/pull/2819/files#diff-ad75e35ac7cb0aad007e6d9e495bbae8R127

Maybe you can make it static and public, outside of the inner class, to be reused from the outsilde-